packages/artifact/src/internal/download-http-client.ts

```diff
@@ -124,66 +194,117 @@ export class DownloadHttpClient {
+        await this.pipeResponseToFile(
           response,
-          stream,
+          destinationStream,
           isGzip(response.message.headers)
         )
+        return
+      } else if (isThrottledStatusCode(response.message.statusCode)) {
+        info(
+          'A 429 response code has been received when attempting to download an artifact'
+        )
+
+        const retryAfterValue = tryGetRetryAfterValueTimeInMilliseconds(
+          response.message.headers
+        )
+        if (retryAfterValue) {
+          incrementAndCheckRetryLimit()
+          await backOffUsingRetryValue(retryAfterValue)
+        } else {
+          // no retry time available, differ to standard exponential backoff
+          incrementAndCheckRetryLimit(response)
+          await backoffExponentially()
+        }
+      } else if (isRetryableStatusCode(response.message.statusCode)) {
+        incrementAndCheckRetryLimit(response)
+        await backoffExponentially()
       } else {
+        // Some unexpected response code, fail immediately and stop the download
         // eslint-disable-next-line no-console
-        console.log(retryResponse)
-        throw new Error(`Unable to download ${artifactLocation}`)
+        console.log(response)
+        throw new Error(
+          `###ERROR### Unable to download ${artifactLocation} ###`
+        )
       }
-    } else {
-      // eslint-disable-next-line no-console
-      console.log(response)
-      throw new Error(`Unable to download ${artifactLocation}`)
     }
   }
 
   /**
-   * Pipes the response from downloading an individual file to the appropriate stream
-   * @param response the http response recieved when downloading a file
-   * @param stream the stream where the file should be written to
-   * @param isGzip does the response need to be be uncompressed
+   * Pipes the response from downloading an individual file to the appropriate destination stream while decoding gzip content if necessary
+   * @param response the http response received when downloading a file
+   * @param destinationStream the stream where the file should be written to
+   * @param isGzip a boolean denoting if the content is compressed using gzip and if we need to decode it
    */
-  private async pipeResponseToStream(
+  private async pipeResponseToFile(
     response: IHttpClientResponse,
-    stream: NodeJS.WritableStream,
+    destinationStream: NodeJS.WritableStream,
     isGzip: boolean
   ): Promise<void> {
-    return new Promise(resolve => {
+    await new Promise((resolve, reject) => {
       if (isGzip) {
-        // pipe the response into gunzip to decompress
         const gunzip = zlib.createGunzip()
         response.message
           .pipe(gunzip)
-          .pipe(stream)
+          .pipe(destinationStream)
           .on('close', () => {
             resolve()
           })
+          .on('error', error => {
+            info(
```

Comment:
If you want to add an info before we throw the error, consider adding some more information to it so it's not a static sentence.  If it shows which file failed it's more useful, otherwise the error below would be enough?

Reply:
I added some extra information here. The raw error might be something ambiguous related to gzip or the file system so to the end user they might have no idea what is going.

An extra `info(..)` can give users at least some extra clarity

Reply:
If the error doesn't give enough information to the user, then consider throw a new user facing error, and log the existing error for us to debug.  

At the high level, errors should be indicative of the issue, and best give the user some way to move forward themselves.